Resolve snapshots

plexus opened the pull request:
This still has some issues, do not merge (yet).

Adds an extra flag to freeze/resolve snapshot versions, so they get replaced with immutable references.

I started by refactoring the update code, pulling all zipper related stuff into its own namespace, and adding some more high level functions, e.g. `(transform-coords loc f)` lets you update all coordinate maps by applying a function over them. This simplified the update code quite a bit, and made implementing the new functionality much easier.
 
It also has the benefit that now all version lookups are parallelized in one go, before each `:deps`/`:override-deps`/`:extra-deps`/`:aliases` was still handled separately, one at a time.

This does break some existing behavior which I still need to fix

- [x] formatting in coordinate maps is not preserved
- [x] `^:depot/ignore` currently only works on the artifact symbol, not on the coords map or any of the parents

Still it's already quite useful, to try it out

```
clj -Sdeps '{:deps {olical/depot {:git/url "https://github.com/plexus/depot.git" :sha "706532e554daf038bc17c173e5db40ba075d11c3"}}}' -m depot.outdated.main --resolve-snapshots
```

plexus committed:
Factor the zipper helpers into their own namespace.

Add our own version of map-vals (similar to how we already had map-keys), to
take uneval nodes into account.

Update Kaocha and enable the notifier.

plexus committed:
Implement generic zipper transformations and refactor --update.

All incidental complexity of dealing with zippers has been moved to a separate
namespace, which now contains high level functions for iterating over and
transforming artifact coordinates in a deps.edn file.

plexus committed:
Add a feature to "freeze" snapshot versions

--resolve-snashots will replace any -SNAPSHOT version with its
"real" (immutable, timestamped) version.

plexus committed:
Mention --resolve-snapshots in the README

puredanger commented on README.md:
```diff
@@ -68,6 +68,18 @@ can instead pass one or more filenames in explicitly.
 
+## Freezing snapshots
+
+Maven versions ending in `-SNAPSHOT` are mutable, a maintainer can publish as many snapshots as they like, all with the same version string. This means that re-running the same code twice might yield different results, if in the meanwhile a new snapshot was released.
```
Rather than "mutable" versions, Maven refers to things like SNAPSHOT (and RELEASE and LATEST) as "virtual" versions, in case you want to match their terminology. I think that term is better at conveying that they are not mutable, but rather a point of indirection that must be resolved (kind of analogous to Clojure atoms - identities that point to a stream of immutable values).

plexus replied:
Thanks Alex, I'll update the wording. Maybe the flag itself could be `--resolve-virtual` and also take RELEASE and LATEST into account.

plexus committed:
Fix issues with resolve-snapshots, rename to resolve-virtual

Preserver formatting in coords maps, and correctly honor :depot/ignore metadata.
This also fixes similar issues in the --update code, introduced by the earlier
zipper refactoring.

plexus commented:
I've addressed the issues I mentioned before, and renamed the flag to `--resolve-virtual`, plus updated the phrasing in the README as @puredanger suggested.

To try this out add one or more SNAPSHOT/RELEASE/LATEST versions in your deps.edn, and run

```
clj -Sdeps '{:deps {olical/depot {:git/url "https://github.com/plexus/depot.git" :sha "4d93765ca35dab768e331dfbc02a366f33239acf"}}}' -m depot.outdated.main --resolve-virtual
```

@Olical wdyt? Is this something you're interested to have in the main branch?

Olical commented:
Sweeeet this looks great, thanks a lot! I'll give it a review but I'll probably end up giving it a merge and deploy if it works okay for me :smile:

Olical commented on src/depot/outdated/resolve_virtual.clj:
```diff
@@ -0,0 +1,50 @@
+(ns depot.outdated.resolve-virtual
+  (:require [clojure.tools.deps.alpha.reader :as reader]
+            [depot.zip :as dzip]
```
I love that you've moved this into a zip ns, makes a lot of sense!

Olical reviewed (approved):
Yep, honestly LGTM, I was considering recommending that `[ERROR]` went to stderr but the more I think about it the more on the fence I am so it's fine as it is. Thanks a lot! Another wonderful feature!

Olical commented:
Released under `v1.8.0`, thank you!

Considering AOT compiling before deploy in the future :thinking:

plexus commented:
Reading over the diff again there are a few things I'd still like to tidy
up, plus an actual bug that I didn't spot before. I'll open a follow up PR.

On Sat, Mar 23, 2019, 17:31 Oliver Caldwell <notifications@github.com>
wrote:

> Released under v1.8.0, thank you!
>
> Considering AOT compiling before deploy in the future 🤔
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Olical/depot/pull/17#issuecomment-475883944>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAB91BvhBDwgYDY22uaPeiFO18CaA4gIks5vZlbUgaJpZM4btd2i>
> .
>

Olical commented:
Oopsie 😬 I was clearly not paying enough attention. I'll get any fixes out
asap.

On Sat, 23 Mar 2019, 16:52 Arne Brasseur, <notifications@github.com> wrote:

> Reading over the diff again there are a few things I'd still like to tidy
> up, plus an actual bug that I didn't spot before. I'll open a follow up PR.
>
> On Sat, Mar 23, 2019, 17:31 Oliver Caldwell <notifications@github.com>
> wrote:
>
> > Released under v1.8.0, thank you!
> >
> > Considering AOT compiling before deploy in the future 🤔
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Olical/depot/pull/17#issuecomment-475883944>, or
> mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAB91BvhBDwgYDY22uaPeiFO18CaA4gIks5vZlbUgaJpZM4btd2i
> >
> > .
> >
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/Olical/depot/pull/17#issuecomment-475885823>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AATPXewsFpmobGkLWCD3F1CPWdiNAqNUks5vZlvlgaJpZM4btd2i>
> .
>

plexus commented:
My bad, I honestly also thought I was done here :)
